### ISSUE TITLE
Sources are off by 1 sample.

### ISSUE THREAD

**Issue by pseeth** [justinsalamon/scaper#73]
"Every once in a blue moon, one of the dumped sources will missing 1 sample.   This is not a show-stopper, since our work-around is to check them and pad them in our subsequent scripts, but might be good to find the bug in the end."

There is a commented out test case for this already, but a fix upstream may useful to make sure everything is the same length so that we can uncomment out this test case. I believe this issue originates from sox.

**Comment by epicycles** [comment]
To give an idea of the frequency, below are the cases from a batch of 1000. 
This is using PR55, and most cases are with the sources, but sometimes (e.g., in example607) it also happens on the mixture.   I do not yet know if the missing / extra samples are from the beginning or end of the file.  If a sample is ever missing/added at the beginning, and specifically with reference to PR 55, then there could be a time shift between sources and mixture which would be rather troublesome, since then sources won't add up to the mixture.   Current workaround is to pad/trim the sources to the same length and then reconstitute the mixture from the sources.

```
In example: 35/1000
Warning: Scaper output on eval/example0034_sources/foreground/1.wav is 159999 samples instead of 160000
In example: 607/1000
Warning: Scaper output on eval/example0606.wav is 160001 samples instead of 160000
In example: 607/1000
Warning: Scaper output on eval/example0606_sources/foreground/0.wav is 160001 samples instead of 160000
In example: 881/1000
Warning: Scaper output on eval/example0880_sources/foreground/0.wav is 159999 samples instead of 160000

```

**Comment by justinsalamon** [comment]
Related: #74

**Comment by justinsalamon** [comment]
@pseeth @epicycles would it make sense to add the "check-and-pad" functionality directly into Scaper (e.g. via #55), while this issue persists? Or would you rather leave the issue to the end user to handle? At the very least, we could check and issue a scaper warning.

If it's indeed a sox issue, we may not be able to fix this other than by implementing #76 (which is a tall order for now).

**Comment by pseeth** [comment]
This really appears to be a bizarre sox issue. We trim files using sox but using seconds which seems to cause rounding errors. One easy, clean way to fix this is to use pysoundfile, but that's a requirement only in the tests, not in the main library. Thoughts on making soundfile a full requirement? That would make it very easy to fix and would be needed to implement #76 eventually anyway.

**Comment by pseeth** [comment]
Or perhaps, even librosa as a requirement. That might be a bit heavy though. But how many people who use scaper don't have librosa?

**Comment by justinsalamon** [comment]
Cheers for the input @pseeth. In an effort to keep deps as light as possible, I think as long as we don't need most of the functionality offered by librosa we should avoid a librosa dep. 

If pysoundfile does the trick, let's go with that. We can, for now, have an ad-hoc function to fix off-by-one issues caused by sox (e.g. `match_sample_length(length_in_samples)`).

Down the line, when the time comes to completely get rid of sox, I believe we should be able to get pretty much everything done through a combination of pysoundfile and pyrubberband.

Thanks!

**Comment by justinsalamon** [comment]
This issue was supposed to be addressed in #55 but per @epicycles it's still there:

"Relatedly, there also seems to still be an off-by-one error, where sometimes the mixture files come out +1 samples, which is still passed through because the mixture is assumed to be correct.   @haerdogan is updating his PR (#86) to cover that as well."

@pseeth and @haerdogan can please lets discuss the problem/fix here and once agreed upon decide who will implement and who will CR. Please remember if this bug passed the current tests it means we need at least one new unit test, and must maintain 100% coverage.

Thanks!

**Comment by haerdogan** [comment]
Yeah I tried to fix it, but my fix is causing some test failures. I think if there is trimming going on, then self.duration is not correct anymore.

**Comment by haerdogan** [comment]
I was trying to enforce self.duration on the outputs. The original one was taking the mixed signal's length as the desired duration but the length of the mixed signal can also be off by one once in a while which causes problems. If there is no trim, my solution would work I think. Is there a way to get the duration after trim?

**Comment by pseeth** [comment]
Yeah, I ran into the same thing, thus the fix that I put in is dependent on the mixture length. If that's wrong, then everything is off by 1 sample. But at least everything is off by 1 sample (the mixture as well as all of the sources). Will have to investigate more.

**Comment by pseeth** [comment]
Not entirely ready to PR this but I did look into it and found the following fix. So the current version does this:

https://github.com/justinsalamon/scaper/blob/a3ca80654985ef553aa33737b9cf2a6fb963ae72/scaper/core.py#L1879-L1881

We use the length of the mixture to make sure  all the sources have the same length. This shouldn't have any problems. The issue crops up when the mixture length does not match the requested duration, which I can't find actually being tested for explicitly. This is also really hard to track down as there is so much randomness. The simple fix, however, is to just use `self.duration` to calculate what to fix all of the lengths to. That'd be like this:

```
                duration_in_samples = int(self.duration * self.sr)
                for _audio_file in [audio_path] + isolated_events_audio_path:
                    match_sample_length(_audio_file, duration_in_samples)
```

This works great, as @haerdogan pointed out. I added an assert to make sure that the soundscape duration is as expected in the tests to make sure.  But this solution breaks when trimming is applied? Why?  Looking into that more, it occurs when `generate_from_jams` is used after a `scaper.trim` call. Here are the relevant lines in `generate_from_jams`:

https://github.com/justinsalamon/scaper/blob/a3ca80654985ef553aa33737b9cf2a6fb963ae72/scaper/core.py#L147-L168

So far so good, we look up the duration in the sandbox and give to the Scaper object. Then, if the fix above is applied, we would use `self.duration * self.sr` to calculate the correct length of the mixture. But if `scaper.trim` is applied right before the call to `generate_from_jams`, here's what happens to duration:

https://github.com/justinsalamon/scaper/blob/a3ca80654985ef553aa33737b9cf2a6fb963ae72/scaper/core.py#L246-L250

The jams duration gets modified by the `jams.slice` call above this. Then the sandbox duration, which is used in `generate_from_jams` uses that duration to create the Soundscape. But the logic in `generate_from_jams` has always been that the audio is generated in its entirety and THEN trimmed if there are slice ops involved. This worked great in the past because `_generate_audio` had never accessed `self.duration` when building the audio. Now that it does though, it uses the trimmed duration rather than the original duration. If the lines above when trimming change to this:

```
            # Update specs in sandbox
            ann.sandbox.scaper['n_events'] = n_events
            ann.sandbox.scaper['polyphony_max'] = poly
            ann.sandbox.scaper['polyphony_gini'] = gini
            # ann.sandbox.scaper['duration'] = ann.duration # DON'T DO THIS?
```

Then everything passes as expected. The durations all match because self.duration is what it's supposed to be, and the trims also all work. Will need more work before I'm ready to PR this though as this is a hack but I thought I'd share discoveries, here!

**Comment by pseeth** [comment]
As an aside, most of these bugs stem from how `sox` requires duration in seconds instead of samples when doing operations like this:

https://github.com/justinsalamon/scaper/blob/a3ca80654985ef553aa33737b9cf2a6fb963ae72/scaper/core.py#L1801-L1812

**Comment by justinsalamon** [comment]
Thanks for the detailed analysis @pseeth.

Changing the following lines to use `self.duration` instead of the mix duration for ensuring all sample counts match sounds like a good solution:
https://github.com/justinsalamon/scaper/blob/a3ca80654985ef553aa33737b9cf2a6fb963ae72/scaper/core.py#L1879-L1881

But, there's the remaining issue with `_generate_from_jams` you've mentioned.

I'm concerned I don't remember 100% how the logic works everywhere to say with confidence that commenting out `ann.sandbox.scaper['duration'] = ann.duration` doesn't cause any issues anywhere.

I'm a little weirded out by the statement that _generate_from_jams uses the post-trim duration, but then still applies the trims, and somehow it all works out. I'll have to dig into the code to gain more clarity.

**Comment by justinsalamon** [comment]
Documenting findings as I go along.

* `scaper.trim()` invokes `jams.slice()`, which in turn invokes `jams.trim()` and adjusts the event start times in the trimmed annotation so they are all relative to the new start time
* `scaper.trim()` first changes the jam file using `jams.slice()`. It then updates the following:
https://github.com/justinsalamon/scaper/blob/9f48f0aaf685a0bc20fa7a527903721d38da2c1d/scaper/core.py#L246-L250
* the trim operation will also update the duration value in the following places:
    * `jam.file_metadata.duration`
    * `jam.ann.duration`
    * `jam.ann.sandbox.scaper.duration`
* This means all fields storing duration are updated to the trimmed duration
* `scaper.trim` does NOT update any of the fields in the value dict of event observations. This means `obs.value.event_duration` and `obs.value.event_time` maintain their original values prior to the trim.
* Each call to `scaper.trim` is documented in `jam.ann.sandbox` with one `trim` and one `slice` dictionaries. 
* Each call to `scaper.trim`, which invokes `jams.slice`, is also documented in the jam sandbox `jam.sandbox` with one `slice` dictionary (note, unlike the scaper sandbox, this does not add a `trim` dictionary to the jam sandbox)

**Comment by justinsalamon** [comment]
Findings about `generate_from_jams()`:

* It obtains the soundscape duration from `ann.sandbox.scaper.duration`:
https://github.com/justinsalamon/scaper/blob/9f48f0aaf685a0bc20fa7a527903721d38da2c1d/scaper/core.py#L147-L150
* It sets all the synthesis parameters from the scaper sandbox and copies the entire sandbox over:
https://github.com/justinsalamon/scaper/blob/9f48f0aaf685a0bc20fa7a527903721d38da2c1d/scaper/core.py#L152-L164
* It then generates the audio from the jams annotation:
https://github.com/justinsalamon/scaper/blob/9f48f0aaf685a0bc20fa7a527903721d38da2c1d/scaper/core.py#L165-L168

Here's where it gets tricky:
* `_generate_audio` uses `self.duration` to set the soundscape duration. As noted above, `self.duration` in this case is determined by the value in `ann.sandbox.scaper.duration`, which is the updated trimmed duration
* `_generate_audio` refers to the following fields in the `value` dict of each observation:
    * `e.value['role']`
    * `e.value['source_file']`
    * `e.value['source_time']`
    * **`e.value['event_duration']`**
    * `e.value['pitch_shift']`
    * `e.value['time_stretch']`
    * **`e.value['event_time']`**
    * `e.value['label']`
* The values in bold above, which remained unmodified from the original soundscape prior to trimming, are used as is to create the isolated events and the mix. This could, in principle, cause weird interactions since these values are unadjusted, and yet `self.duration` is adjusted to the trimmed duration.
* `self.duration` is used in the following places in `_generate_audio()`:
https://github.com/justinsalamon/scaper/blob/9f48f0aaf685a0bc20fa7a527903721d38da2c1d/scaper/core.py#L1694-L1706
https://github.com/justinsalamon/scaper/blob/9f48f0aaf685a0bc20fa7a527903721d38da2c1d/scaper/core.py#L1799-L1812

The first block uses `self.duration` to determine how many times to tile the background to itself. The second block uses `self.duration` to determine by how much to pre/post pad foreground events with silence so that their duration matches that of the complete soundscape. Both of these operations could potentially lead to problems, because `e.value` contains all the old pre-trim values, whereas `self.duration` is the new trimmed duration. **THIS IS A BUG**.

I believe the possible negative effects of the bug could include (not comprehensive):
* not tiling the background to itself correctly (though this feature is broken right now #70)
* not adding any postpadding to foreground events prior to trimming them

Somehow, this bug has miraculously passed all tests. I believe it's due to the following:
* Next in the process, `_generate_from_jams` will look for `slice` operations in the annotation `ann.sandbox`:
https://github.com/justinsalamon/scaper/blob/9f48f0aaf685a0bc20fa7a527903721d38da2c1d/scaper/core.py#L172
* For every slice operation, it will apply a sox `tfm.trim` operation to the soundscape audio file and all isolated events:
https://github.com/justinsalamon/scaper/blob/9f48f0aaf685a0bc20fa7a527903721d38da2c1d/scaper/core.py#L173-L187
* This means that foreground events that were missing post-padding get trimmed such that the post-padding doesn't impact anything
* If the background is longer than the original (and trimmed) soundscape durations, I don't believe any negative impacts are observed
* If however the background were shorted than the original soundscape and/or shorter than the trimmed soundscape, I believe we'd observe some incorrect behavior (e.g. the foreground events are where they should be in terms of start times, back the background isn't).

**Comment by justinsalamon** [comment]
All this analysis basically leads us to @pseeth 's proposed solution: make sure `self.duration` in the scaper object used in `generate_from_jams()` is set to the original (pre-trim) duration of the soundscape. From there everything else should work correctly.

Things we need to do for a complete fix:
* Determine which field in the JAMS file should store the original soundscape duration. This could be one of the existing fields (e.g. `ann.sandbox.scaper.duration` as @pseeth has proposed), or a new field.
* If an existing field is used for this, must make sure it's not used anywhere else that could cause bugs
* Add a test that can catch this issue.

Now reviewing code to determine my thoughts re: the above points.

**Comment by pseeth** [comment]
I think both the original duration and the trimmed duration should be kept. If no trim was applied, then original duration and trimmed duration should be the same. If applied, then we use the original duration. I see that in `generate_from_jams` that a trim operation is looked for by looking for `slice` in ops. Should we instead make that a field in the sandbox that explicitly says this is a trimmed soundscape?

**Comment by justinsalamon** [comment]
Right now scaper.trim operations are documented in the "trim" dictionary in `ann.sandbox`. Note that it's not a binary "trimmed or not", because one can apply multiple consecutive trims. So instead it's a dict that points to a list of trim operations. If the dict key isn't even there it means there are no trim operations. If the dict key is there, then the list length indicates how many trim operations were performed.

As such, I don't know that there's need to add more info into `ann.sandbox.scaper` about the trimming, at least for now.

I do however tend to agree we should have a new field for documenting the original soundscape duration.

**Comment by pseeth** [comment]
Okay to understand the fix in totality:

1. In `ann.sandbox.scaper`, we will have two keys: `duration` (which we have now) and something like `original_duration`. When the soundscape is generated for the first time, the values in each of these will be the same.
2. `trim` will modify `duration`, but `original_duration` will always be used in `generate_from_jams`, as the audio gets trimmed after being generated.
3. `_generate_audio` will use `self.duration` when passing the audio through `match_sample_length`.

Did I miss anything?

On the tests side, we need a test to ensure that the mixture duration is exactly what we expect. I'll find a jams file that fails this test under Scaper 1.3.3 reliably.

**Comment by justinsalamon** [comment]
@pseeth this sounds perfect. Let's get that PR up!

**Comment by justinsalamon** [comment]
For documentation, here is how I see the current bug causing a failure:
* Original soundscape duration is 10s
* You use a BG source file that's only 3s long
* You add an FG event between 4-6s
* BG should get tiled, given you a soundscape with BG everywhere and FG event between 4-6s

* Now you trim the soundscape to the content between 4-6s
* The new soundscape should be 2s long, with the FG event intact and BG in the background (from the 2nd tile of the BG source file)

* Now call `generate_from_jams()` using the JAMS file of the trimmed soundscape
* Since the new duration is 2s, the BG won't get tiled at all, i.e. it will be 3s of sound followed by silence
* The FG event will get prepadding based on its original start time, meaning there will be 4s of silence and then 2s of sound. There will be no post padding (since soundscape_duration (=2) - event_duration (=2) equals 0)
* The FG and BG will get mixed together, resulting in 2 (or 3, not sure) seconds of BG sound, following by silence until 4s, and then the FG event for 2s
* Then the mix will get trimmed to the content between 4-6
* The results is a 2s soundscape that maintains the FG event (correct), but doesn't have any sound in the BG (incorrect)

Testing for this failure case would require replicating the above (create 10s soundscape, trim, generate from jams). Right now the tiling functionality is broken (cf. #70).

**Comment by justinsalamon** [comment]
The **TWO** bugs discussed in this issue (off by 1 and incorrect soundscape duration when calling generate_from_jams) have been addressed by #90, hurrah!
